src/neo4j_genai/components/entity_relation_extractor.py

```diff
@@ -0,0 +1,108 @@
+from neo4j_genai.llm import LLMInterface
+from neo4j_genai.pipeline.component import Component
+
+logger = logging.getLogger(__name__)
+
+
+class EntityRelationExtractor(Component):
+    async def run(self, chunks: TextChunks, **kwargs: Any) -> Neo4jGraph:
+        # for each chunk, returns a dict with entities and relations keys
+        return Neo4jGraph(nodes=[], relationships=[])
+
+
+class OnError(enum.Enum):
+    RAISE = "RAISE"
+    IGNORE = "CONTINUE"
+
+
+class LLMEntityRelationExtractor(EntityRelationExtractor):
+    def __init__(
+        self,
+        llm: LLMInterface,
+        prompt_template: ERExtractionTemplate | str = ERExtractionTemplate(),
+        on_error: OnError = OnError.RAISE,
+    ) -> None:
+        self.llm = llm  # with response_format={ "type": "json_object" },
+        if isinstance(prompt_template, str):
+            template = PromptTemplate(prompt_template, expected_inputs=[])
+        else:
+            template = prompt_template
+        self.prompt_template = template
+        self.on_error = on_error
+
+    def update_ids(self, chunk_index: int, graph: Neo4jGraph) -> Neo4jGraph:
+        for node in graph.nodes:
+            node.id = f"{chunk_index}:{node.id}"
+        for rel in graph.relationships:
+            rel.start_node_id = f"{chunk_index}:{rel.start_node_id}"
+            rel.end_node_id = f"{chunk_index}:{rel.end_node_id}"
+        return graph
+
+    # TODO: fix the type of "schema" and "examples"
+    @validate_call
+    async def run(
+        self,
+        chunks: TextChunks,
+        schema: Union[BaseModel, Dict[str, Any], None] = None,
+        examples: Any = None,
+        **kwargs: Any,
+    ) -> Neo4jGraph:
+        schema = schema or {}
+        examples = examples or ""
+        graph = Neo4jGraph()
+        for index, chunk in enumerate(chunks.chunks):
+            prompt = self.prompt_template.format(
+                text=chunk.text, schema=schema, examples=examples
+            )
+            llm_result = self.llm.invoke(prompt)
+            try:
+                result = json.loads(llm_result.content)
+            except json.JSONDecodeError:
+                if self.on_error == OnError.RAISE:
+                    logger.error(f"LLM response is not valid JSON {llm_result.content}")
+                    raise LLMGenerationError(
+                        f"LLM response is not valid JSON {llm_result.content}"
+                    )
+                result = {"nodes": [], "relationships": []}
+            try:
+                chunk_graph = Neo4jGraph(**result)
+            except ValidationError:
+                if self.on_error == OnError.RAISE:
+                    logger.error(
+                        f"LLM response has improper format {llm_result.content}"
+                    )
+                    raise LLMGenerationError(
+                        f"LLM response has improper format {llm_result.content}"
+                    )
+                chunk_graph = Neo4jGraph()
```

Comment:
So if the user chooses to set `on_error` to `IGNORE`, `chunk_graph` will be empty. Do you think is likely to happen? What should a user do in this situation? Try different prompting or LLM parameters?

Reply:
At the moment, there is no way to resume a failed pipeline, so I think it could make sense to ignore errors for a specific chunk, otherwise one failing chunk out of 100 and you'll have to rerun everything. But probably we could implement some retry mechanism in the future to make the user experience better, or make `pipeline.run` method idemptent somehow.